engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/ISBNValidator.java

```diff
@@ -54,37 +43,101 @@ public boolean isValid(CharSequence isbn, ConstraintValidatorContext context) {
 		String digits = NOT_DIGITS_OR_NOT_X.matcher( isbn ).replaceAll( "" );
 
 		// Check if the length of resulting string matches the expecting one
-		if ( digits.length() != length ) {
+		if ( !isbnValidationAlgorithm.isValidLength( digits.length() ) ) {
 			return false;
 		}
 
-		return checkChecksumFunction.apply( digits );
+		return isbnValidationAlgorithm.isValidChecksum( digits );
 	}
 
-	/**
-	 * Check the digits for ISBN 10 using algorithm from
-	 * <a href="https://en.wikipedia.org/wiki/International_Standard_Book_Number#ISBN-10_check_digits">Wikipedia</a>.
-	 */
-	private boolean checkChecksumISBN10(String isbn) {
-		int sum = 0;
-		for ( int i = 0; i < isbn.length() - 1; i++ ) {
-			sum += ( isbn.charAt( i ) - '0' ) * ( 10 - i );
+	private interface ISBNValidationAlgorithm {
+		boolean isValidLength(int length);
+
+		boolean isValidChecksum(String isbn);
+
+		static ISBNValidationAlgorithmImpl from(ISBN.Type type) {
+			Contracts.assertNotNull( type );
+			switch ( type ) {
+				case ISBN_10:
+					return ISBNValidationAlgorithmImpl.ISBN_10;
+				case ISBN_13:
+					return ISBNValidationAlgorithmImpl.ISBN_13;
+				case ANY:
+				default:
+					return ISBNValidationAlgorithmImpl.ANY;
+			}
 		}
-		sum += isbn.charAt( 9 ) == 'X' ? 10 : isbn.charAt( 9 ) - '0';
-
-		return ( sum % 11 ) == 0;
 	}
 
-	/**
-	 * Check the digits for ISBN 13 using algorithm from
-	 * <a href="https://en.wikipedia.org/wiki/International_Standard_Book_Number#ISBN-13_check_digit_calculation">Wikipedia</a>.
-	 */
-	private boolean checkChecksumISBN13(String isbn) {
-		int sum = 0;
-		for ( int i = 0; i < isbn.length(); i++ ) {
-			sum += ( isbn.charAt( i ) - '0' ) * ( i % 2 == 0 ? 1 : 3 );
+	private enum ISBNValidationAlgorithmImpl implements ISBNValidationAlgorithm {
+
+		ISBN_10 {
+			@Override
+			public boolean isValidChecksum(String isbn) {
+				return checkChecksumISBN10( isbn );
+			}
+
+			@Override
+			public boolean isValidLength(int length) {
+				return 10 == length;
+			}
+		},
+		ISBN_13 {
+			@Override
+			public boolean isValidChecksum(String isbn) {
+				return checkChecksumISBN13( isbn );
+			}
+
+			@Override
+			public boolean isValidLength(int length) {
+				return 13 == length;
+			}
+		},
+		ANY {
+			@Override
+			public boolean isValidLength(int length) {
+				return 10 == length || 13 == length;
+			}
+
+			@Override
+			public boolean isValidChecksum(String isbn) {
+				int length = isbn.length();
+				if ( length == 10 ) {
+					return checkChecksumISBN10( isbn );
+				}
+				else if ( length == 13 ) {
+					return checkChecksumISBN13( isbn );
+				}
+				throw new IllegalStateException( "Invalid/unsupported isbn value length" );
+			}
+		},
```

Comment:
I would remove the `,`. I know modern JVM doesn't have a problem with it but I'm not a big fan of it.